Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/ProgressIndicators/ProgressIndicatorOrbsRotator.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Threading.Tasks;
+using TMPro;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.SceneTransitions
+{
+    /// <summary>
+    /// This class manages the 'rotating circle of dots' effect
+    /// that is used as a Progress Indicator effect.
+    /// </summary>
+    public class ProgressIndicatorOrbsRotator : MonoBehaviour, IProgressIndicator
+    {
+        public Transform MainTransform { get { return transform; } }
+        public ProgressIndicatorState State { get { return state; } }
+        public float Progress { set { progress = value; } }
+        public string Message { set { messageText.text = value; } }
+
+        [SerializeField]
+        private GameObject[] orbs = null;
+        
+        [SerializeField]
+        private TextMeshPro messageText = null;
+
+        [SerializeField]
+        private ProgressIndicatorState state = ProgressIndicatorState.Closed;
+        
+        public float RotationSpeedRawDegrees;
+        public float SpacingDegrees;
+        public float Acceleration;
+        public int Revolutions;
+        public bool TestStop = false;
+        public bool HasAnimationFinished = false;
+
+        private float progress;
+        private float timeElapsed;
+        private int deployedCount;
+        private bool timeUpdated;
+        private float[] angles;
+        private float timeSlice;
+        private float deg2rad = Mathf.PI / 180.0f;
+        private Renderer[] dots = null;
+        private bool stopRequested;
+        private float rotationWhenStopped;
+        private MaterialPropertyBlock[] propertyBlocks = null;
+
+        public async Task OpenAsync()
+        {
+            gameObject.SetActive(true);
+
+            state = ProgressIndicatorState.Opening;
+
+            StartOrbs();
+
+            await Task.Yield();
+
+            state = ProgressIndicatorState.Open;
+        }
+
+        public async Task CloseAsync()
+        {
+            state = ProgressIndicatorState.Closing;
+
+            StopOrbs();
+
+            while (!HasAnimationFinished)
+            {
+                await Task.Yield();
+            }
+
+            state = ProgressIndicatorState.Closed;
+
+            gameObject.SetActive(false);
+        }
+
+        private void StartOrbs()
+        {
+            rotationWhenStopped = 0.0f;
+            stopRequested = false;
+            timeSlice = 0;
+            timeUpdated = false;
+            deployedCount = 1;
+            timeElapsed = 0.0f;
+            angles = new float[orbs.Length];
```

Comment:
Nit: Consider only doing this if null or if angles.length != orbs.Length. Otherwise there's no reason to blow away a chunk of memory that's identical to what is needed here.